### PR TITLE
feat: content highlights stepper updates, including refactor to use context selectors and connects to Algolia for course search results

### PR DESCRIPTION
**Pull request by brobro10000**
### v2

* Refactors to use `use-context-selector` for content highlights, which significantly reduces the performance impacts on component re-rendering when using our architectural approach to using vanilla React context; acts as a polyfill since official support for context selectors is still a proposal).
  * Note: This ultimately makes the reducer previously implemented unnecessary in favor of a more streamlined, state-like way to manage state than dispatching reducer actions.
  * Readings:
    * https://github.com/dai-shi/use-context-selector
    * https://github.com/reactjs/rfcs/pull/119
* Continues on with the title step validation logic. Adds a validation message to the `Stepper` step description.
* Persists data between the `Stepper` steps.
* Disables `DataTable` row selection when the cap is reached.
* Connects the "Select content" and "Confirm and and publish" steps with Algolia.
  * _Note: requires configuring Algolia API environment variables. I'm using the production Algolia index for now (since stage course data is janky) with a hardcoded (test) production enterprise customer UUID. Recommended to use `.env.private` to set these environment variables for local development._
* Sends `POST` to `/api/v/1/highlight-sets-admin/` on final CTA click from "Confirm and publish" step to create highlight, prepending it to any existing highlight sets in the `EnterpriseAppContext.enterpriseCuration` data without needing to refresh from API.
* Depends on upstream Paragon changes:
  * https://github.com/openedx/paragon/pull/1804
* Depends on the Algolia index having `aggregation_key` configured as an attribute for faceting (**PR incoming**).

### v1

Added Context, and refactored test/code to pass with context values for title and stepper logic
Initial work on data persist, more work is needed on it at this point
Set up reducer for persisting stepper data
Added logic to title to reflect limit on length of highlight title
Added additional UI changes to match language of the figma 
Added reducer for 'Last deleted item' to be used in toast on deletion

# For all changes

- [ ] Ensure adequate tests are in place (or reviewed existing tests cover changes)

# Only if submitting a visual change

- [ ] Ensure to attach screenshots
- [ ] Ensure to have UX team confirm screenshots

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -68,9 +67,11 @@
     "redux-mock-store": "1.5.4",
     "redux-thunk": "2.3.0",
     "regenerator-runtime": "0.13.7",
+    "scheduler": "^0.23.0",
```
Comment (adamstankiewicz) [inline package.json:70]: required peer dependency of `use-context-selector` below

File: src/components/ContentHighlights/HighlightStepper/HighlightStepperSelectContentDataTable.jsx [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+      <HighlightStepperSelectContentDataTable
+        selectedRowIds={currentSelectedRowIds}
+        onSelectedRowsChanged={setCurrentSelectedRowIds}
+      />
+    </InstantSearch>
+  );
+};
+
+const PriceTableCell = ({ row }) => {
+  const contentPrice = row.original.firstEnrollablePaidSeatPrice;
```
Comment (adamstankiewicz) [inline src/components/ContentHighlights/HighlightStepper/HighlightStepperSelectContentDataTable.jsx:61]: [note] tests will be added for covered in a future PR

File: src/components/ContentHighlights/HighlightStepper/HighlightStepperTitleInput.jsx [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+const HighlightStepperTitleInput = () => {
+  const { setHighlightTitle } = useContentHighlightsContext();
+  const highlightTitle = useContextSelector(ContentHighlightsContext, v => v[0].stepperModal.highlightTitle);
+  const [titleLength, setTitleLength] = useState(highlightTitle?.length || 0);
+  const [isInvalid, setIsInvalid] = useState(false);
+
+  const handleChange = (e) => {
+    if (e.target.value.length > 60) {
+      setIsInvalid(true);
+      setHighlightTitle({
```
Comment (adamstankiewicz) [inline src/components/ContentHighlights/HighlightStepper/HighlightStepperTitleInput.jsx:18]: [note] A test will be added in future PR for this coverage.

File: src/components/ContentHighlights/HighlightStepper/SelectContentSelectionCheckbox.jsx [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+import { MAX_CONTENT_ITEMS_PER_HIGHLIGHT_SET } from '../data/constants';
+import { ContentHighlightsContext } from '../ContentHighlightsContext';
+
+const SelectContentSelectionCheckbox = ({ row }) => {
+  const {
+    indeterminate,
+    checked,
+    ...toggleRowSelectedProps
+  } = row.getToggleRowSelectedProps();
```
Comment (adamstankiewicz) [inline src/components/ContentHighlights/HighlightStepper/SelectContentSelectionCheckbox.jsx:14]: Test coverage will be added in a separate PR here.

File: src/components/ContentHighlights/HighlightStepper/SelectContentSelectionStatus.jsx [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+import { useContextSelector } from 'use-context-selector';
+import PropTypes from 'prop-types';
+import classNames from 'classnames';
+import { Button, DataTableContext } from '@edx/paragon';
+
+import { ContentHighlightsContext } from '../ContentHighlightsContext';
+
+const SelectContentSelectionStatus = ({ className }) => {
+  const { toggleAllRowsSelected } = useContext(DataTableContext);
+  const currentSelectedRowsCount = useContextSelector(
```
Comment (adamstankiewicz) [inline src/components/ContentHighlights/HighlightStepper/SelectContentSelectionStatus.jsx:11]: [note] Test coverage will be added in a future PR here.

File: src/components/ContentHighlights/data/hooks.js [hunk 6]
```diff
@@ -35,4 +28,58 @@ export const useHighlightSetsForCuration = (enterpriseCuration) => {
   }, [enterpriseCuration]);
 
   return highlightSets;
-};
+}
+
+export function useContentHighlightsContext() {
```
Comment (adamstankiewicz) [inline src/components/ContentHighlights/data/hooks.js:33]: nit: We may want to add a JSDOc to this hook to document its intent/purpose, e.g.:

```jsx
/**
 * Defines an interface to mutate the `ContentHighlightsContext` context value.
 */
export function useContentHighlightsContext() {
```

Comment (adamstankiewicz) [reply src/components/ContentHighlights/data/hooks.js:33]: [consideration] I wonder if there's a better name for `useContentHighlightsContext` to reflect this intent/purpose.

File: src/components/ContentHighlights/ContentHighlightRoutes.jsx [hunk 7]
```diff
@@ -2,22 +2,43 @@ import React from 'react';
...
 
 const ContentHighlightRoutes = ({ enterpriseSlug }) => {
   const baseContentHighlightPath = `/${enterpriseSlug}/admin/${ROUTE_NAMES.contentHighlights}`;
   return (
     <>
       <Route
         path={baseContentHighlightPath}
-        component={ContentHighlightsDashboard}
+        render={(routeProps) => (
+          <BaseContentHighlightRoute {...routeProps}>
```
[the remark this answers is not in the document]
Comment (brobro10000) [reply src/components/ContentHighlights/ContentHighlightRoutes.jsx:29]: Leaving open as a reminder to complete

File: src/components/ContentHighlights/ContentHighlightSetCard.jsx [hunk 8]
```diff
@@ -16,16 +17,15 @@ const ContentHighlightSetCard = ({
...
-  } = useContext(ContentHighlightsContext);
+  const { openStepperModal } = useContentHighlightsContext();
   /* Stepper Draft Logic (See Hook) - End */
   const handleHighlightSetClick = () => {
     if (isPublished) {
-      // redirect to individual highlighted courses based on uuid
-      return history.push(`/${enterpriseSlug}/admin/${ROUTE_NAMES.contentHighlights}/${highlightSetUUID}`);
+      // redirect to individual highlighted set based on uuid
+      history.push(`/${enterpriseSlug}/admin/${ROUTE_NAMES.contentHighlights}/${highlightSetUUID}`);
+      return;
```
[the remark this answers is not in the document]
Comment (brobro10000) [reply src/components/ContentHighlights/ContentHighlightSetCard.jsx:26]: Leaving open as a reminder to complete

File: src/components/ContentHighlights/CurrentContentHighlightHeader.jsx [hunk 9]
```diff
@@ -1,40 +1,28 @@
...
-        <Button
-          iconBefore={Add}
-          onClick={handleNewClick}
-        >
-          New
-        </Button>
-      </ActionRow>
-      <p>
-        Create up to 8 highlight collections for your learners.
-      </p>
```
[the remark this answers is not in the document]
Comment (brobro10000) [reply src/components/ContentHighlights/CurrentContentHighlightHeader.jsx:34]: Leaving open as a reminder to complete

File: src/components/ContentHighlights/HighlightStepper/ContentHighlightStepper.jsx [hunk 10]
```diff
@@ -1,108 +1,200 @@
...
+    setIsPublishing(true);
+    try {
+      const newHighlightSet = {
+        title: highlightTitle,
+        isPublished: true,
+        // TODO: pass along the selected content keys!
+      };
+      const response = await EnterpriseCatalogApiService.createHighlightSet(enterpriseId, newHighlightSet);
+      const result = camelCaseObject(response.data);
+      const transformedHighlightSet = {
```
[the remark this answers is not in the document]
Comment (brobro10000) [reply src/components/ContentHighlights/HighlightStepper/ContentHighlightStepper.jsx:76]: Leaving open as a reminder to complete

File: src/components/ContentHighlights/HighlightStepper/HighlightStepperConfirmContent.jsx [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+import SkeletonContentCard from '../SkeletonContentCard';
+
+const prodEnterpriseId = 'e783bb19-277f-479e-9c41-8b0ed31b4060';
+
+const BaseReviewContentSelections = ({
+  searchResults,
+  isSearchStalled,
+}) => {
+  if (isSearchStalled) {
+    return (
```
[the remark this answers is not in the document]
Comment (brobro10000) [reply src/components/ContentHighlights/HighlightStepper/HighlightStepperConfirmContent.jsx:27]: Leaving open as a reminder to complete

File: src/components/ContentHighlights/HighlightStepper/HighlightStepperSelectContentDataTable.jsx (hunk 2 above)
Comment (brobro10000) [reply src/components/ContentHighlights/HighlightStepper/HighlightStepperSelectContentDataTable.jsx:61]: Leaving open as a reminder to complete

File: src/components/ContentHighlights/HighlightStepper/HighlightStepperTitleInput.jsx (hunk 3 above)
Comment (brobro10000) [reply src/components/ContentHighlights/HighlightStepper/HighlightStepperTitleInput.jsx:18]: Leaving open as a reminder to complete

File: src/components/ContentHighlights/HighlightStepper/SelectContentSelectionCheckbox.jsx (hunk 4 above)
Comment (brobro10000) [reply src/components/ContentHighlights/HighlightStepper/SelectContentSelectionCheckbox.jsx:14]: Leaving open as a reminder to complete

File: src/components/ContentHighlights/HighlightStepper/SelectContentSelectionStatus.jsx (hunk 5 above)
Comment (brobro10000) [reply src/components/ContentHighlights/HighlightStepper/SelectContentSelectionStatus.jsx:11]: Leaving open as a reminder to complete
